core/embed/trezorhal/unix/common.c

```diff
@@ -28,14 +28,16 @@
 #include "display.h"
 #include "memzero.h"
 
+extern void __attribute__((noreturn)) main_clean_exit(int);
+
 void __attribute__((noreturn)) trezor_shutdown(void) {
   printf("SHUTDOWN\n");
 
   // Wait some time to let the user see the displayed
   // message before shutting down
   hal_delay(3000);
 
-  exit(3);
+  main_clean_exit(3);
```

Comment:
If `trezor_shutdown` is called before MicroPython is initialized, calling `main_clean_exit()` leads to an exception. It's not a big problem since it's an emulator, but it is still an issue.

I wonder if it would be possible to remove the hardcoded constant (3) from here. Is it really used anywhere?

Reply:
@cepetr i recall the two of us discussing whether we need to call `main_clean_exit` from here, and it causing some sort of problem (maybe recursing into trezor_shutdown, or fatal_error, or something)
do you remember if that's not a problem anymore?

Reply:
> calling `main_clean_exit()` leads to an exception

`main_clean_exit` should check whether micropython has been set up before nlr_raising?

Reply:
> @cepetr i recall the two of us discussing whether we need to call `main_clean_exit` from here, and it causing some sort of problem (maybe recursing into trezor_shutdown, or fatal_error, or something) do you remember if that's not a problem anymore?

Yes, you are right. I thought it was solved, but it wasn't. The problem occurs when calling `panic!()` or something similar from the `render()` routine. The call to `nlr_raise()` in `main_clean_exit()` triggers `trezor_shutdown()` again because the code tries to borrow `RefCell<LayoutObj>`, which is already borrowed.

Reply:
> > calling `main_clean_exit()` leads to an exception
> 
> `main_clean_exit` should check whether micropython has been set up before nlr_raising?

I think so.

Reply:
> The call to `nlr_raise()` in `main_clean_exit()` triggers `trezor_shutdown()` again because the code tries to borrow `RefCell<LayoutObj>`

right.

ISTM we should modify the `exit`s for the SDL event loop, but keep this one a raw `exit()`. it seems appropriate that `trezor_shutdown()`, which is a "hard halt" for the real device, also causes a "hard halt" in the emulator, and we should not expect to recover even for prof.py
(we'll get coverage when the emulator is shut down cleanly via `SDL_QUIT` event)

Reply:
check added in 5475558ff22795b24ec380d4bc5b2ce6900251ab, please take a look